### PR TITLE
feat: ai inference improvements, board recognition, CPU usage fixes, and new backends

### PR DESCRIPTION
**Pull request by Aitai**
First, I want to apologize for the size of this pull request. I started this as a private fork to scratch a few itches, and before I knew it I had a stack of changes I think could benefit more people. Thank you so much for building Kaya, it's a fantastic project and it's been a joy to work on.

I'm happy to split this into smaller PRs, squash commits, or rework anything to make review easier.

---

### AI suggesting illegal ko moves
The neural network policy output wasn't filtering the ko-forbidden vertex before ranking candidate moves. `board._koInfo` is now restored from options before featurization so ko legality is respected in all analysis paths.

### Visits cache over-invalidation
Changing `numVisits` was clearing the entire analysis graph cache. It now only invalidates the current position's cache entry, preserving the rest of the analysis you've already done.

---

### Eliminate constant CPU/GPU usage from CSS animations
Several infinite CSS animations (`pulse-border` on the player indicator, `pulse-analyzing`/`pulse-dot` on action buttons, `pulse-indicator` on the analysis mode dot, `cursor-pulse` on the navigation cursor) were causing continuous compositor repaints even when the app was idle. Replaced with static styles. Remaining infinite animations (spinners, drop zone, shortcut recording) are transient and only active during brief operations.

Also narrowed the player indicator's CSS `transition` property from `all` to only `transform, box-shadow`, and optimized `useGamepads` polling to skip re-renders when the gamepad list is unchanged.

### Featurization optimization
- Compute liberty counts once per group instead of running BFS per stone.
- Use direct `signMap[][]` access instead of `board.get()` in the hot
  featurization loop.
- Skip `validateTensorData` loop when debug mode is disabled.

---

### In-browser ONNX → WebGPU model conversion + GPU IO binding
Users can upload any standard KataGo ONNX model and have it auto-converted for WebGPU entirely in the browser — no Python or external tools needed. The converter makes the batch dimension static and decomposes `Softplus`/`LogSoftmax` into GPU-supported primitive ops. Combined with GPU IO binding, pre-allocated GPU buffers, and graph capture, this gives ~8× throughput improvement for full-game batch analysis (~12 ms/pos vs ~100 ms/pos at the default batch size of 4). Batch size is user-configurable (1–16) to trade off GPU memory vs speed.

### MCTS with virtual loss
Adds Monte Carlo Tree Search with configurable visit count. Uses virtual loss to select up to `batch_size` leaves simultaneously (matching the model's compiled batch dimension) for a single batched GPU evaluation, amortizing GPU↔CPU sync overhead across multiple visits instead of paying it once per visit.

### WebNN backend with auto model conversion
Adds a WebNN execution provider with an in-browser converter that makes dimensions static and decomposes unsupported ops. With `freeDimensionOverrides`, ORT achieves ~100% GPU coverage on the MLGraph partition (vs ~8% without). Benchmarks on an AMD Radeon 8060S show batch=4 gives ~4× per-move throughput (~4 ms/move vs ~14 ms/move at batch=1). The batch size slider applies to both WebGPU and WebNN backends.

### PyTorch ROCm/CUDA sidecar (desktop only)
Adds an optional Python sidecar for GPU inference via PyTorch, achieving 261–328 inf/s in FP16 on tested AMD/NVIDIA hardware. Includes Tauri commands for single-position analysis, batch analysis, and benchmarking. Auto-selected when "native" backend is chosen and a GPU is detected.

### Board image recognition importer
New `@kaya/board-recognition` package and dialog for importing board positions from photos:
- Saturation-based board detection with perspective warp
- Two-phase alignment: drag corner handles → click 4 grid intersections
- Stone calibration mode to correct misclassified stones
- Heavy processing runs in a Web Worker to keep the UI responsive

---

## Housekeeping

- All new UI strings added to all 8 supported languages
- Type-checked (`bun run type-check`) and formatted (`bun run format`)
- Tested on both web and desktop

### CODE REVIEW

**Comment by Aitai** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by hadim** [comment]
Thank you for contributing back to that repo. It's great.

I don't have time myself to work on new feature and my hope was to also bring new developers here that can push Kaya forward.

Usually I prefer having smaller, more focused PR instead of a large one. But let's make an exception here. Just please be sure to open focused and smaller PRs for next time :-D

Given the size, the AI review will be a large one as well. But I won't be too picky as long as you're ready to fix any bugs or regressions once that one is merged.

---

The board recognition is great, and my plan was to train a small CNN pytorch model for this and convert it to ONNX but your approach is already useful.

Thanks for the AI performance improvement; those look impressive. Do you think we should modify the default model I uploaded on HF? Or maybe we don't really need an upload from HF now? My goal is for the average user to get the best and most efficient model by default without having him to configure anything given his device (web, desktop, etc). Do you think it's possible?

The Pytorch script is great for the speed, but it requires a Python installation on the host machine. I am happy to keep it as long as it cannot create any friction for the user without Python (just make sure nothing will crash and that option is hidden in that case).

And thanks for all the other fixes and improvements as well.

File: apps/desktop/src-tauri/src/commands.rs [hunk 1]
```diff
@@ -253,9 +263,196 @@ pub fn onnx_get_provider_preference() -> String {
...
+    }
+    #[cfg(not(target_os = "linux"))]
+    {
+        false
+    }
+}
+
+/// Initialize PyTorch GPU engine with a model file
+#[tauri::command]
+pub async fn pytorch_initialize(model_path: String) -> Result<serde_json::Value, String> {
```
Comment (hadim) [inline apps/desktop/src-tauri/src/commands.rs]: 🔒 **Security: Unsanitized file path**

`model_path` is passed directly to the Python sidecar without validation. The Python script will read this path and also write a modified copy (`_kaya_fixed_*`) to the same directory. This could allow reading arbitrary files on the system.

**Suggested fix** — validate that `model_path` resolves to within the app's expected models directory before passing it through:
```rust
let models_dir = app_handle.path().app_data_dir()?.join("models");
let abs_path = std::fs::canonicalize(&model_path)
    .map_err(|e| format!("Invalid model path: {}", e))?;
if !abs_path.starts_with(&models_dir) {
    return Err("Model must be in the app's models directory".to_string());
}
```

Also applies to `onnx_get_cached_model` and `onnx_delete_cached_model` above — `model_id` is used in path construction (`format!("{}.onnx", model_id)`) without sanitization. A value like `../../etc/passwd` would traverse outside the models directory.

File: apps/desktop/src-tauri/src/pytorch_engine.rs [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+        self.stdin
+            .write_all(b"\n")
+            .map_err(|e| format!("Failed to write newline: {}", e))?;
+        self.stdin
+            .flush()
+            .map_err(|e| format!("Failed to flush: {}", e))?;
+
+        let mut response = String::new();
+        self.stdout
+            .read_line(&mut response)
```
Comment (hadim) [inline apps/desktop/src-tauri/src/pytorch_engine.rs]: ⚠️ **Robustness: No timeout on sidecar I/O**

`read_line()` blocks indefinitely. If the Python process hangs or crashes after producing partial output, the entire Tauri app freezes with no way to recover.

Similarly, in `dispose_engine()` (line 361), if `send_command("dispose")` hangs, `process.kill()` is never reached.

**Suggested fix** — consider adding a timeout mechanism (e.g., using `tokio::time::timeout` around the blocking call), or at minimum ensure `dispose_engine()` calls `process.kill()` before attempting `send_command`:
```rust
pub fn dispose_engine() -> Result<(), String> {
    let mut global = ENGINE.lock().map_err(|e| e.to_string())?;
    if let Some(mut engine) = global.take() {
        let _ = engine.process.kill();
        let _ = engine.process.wait();
    }
    Ok(())
}
```

File: packages/ui/src/components/layout/Header.tsx [hunk 3]
```diff
@@ -122,8 +125,16 @@ export const Header: React.FC<HeaderProps> = ({
   }, [fileName, gameInfo.gameName, gameInfo.playerBlack, gameInfo.playerWhite]);
 
   // File loading
+  const IMAGE_EXTENSIONS = ['.jpg', '.jpeg', '.png', '.webp', '.bmp'];
```
Comment (hadim) [inline packages/ui/src/components/layout/Header.tsx]: 💡 **UX: Board recognition feature is hard to discover**

Currently, the only way to access board recognition is to click "Open" and select an image, or drag-drop an image file. The "Open" button is labeled "Open SGF File" and there's no visual cue that image import is supported.

Would you consider adding a dedicated entry point to make this feature discoverable? For example, a menu item in the "More actions" (⋯) dropdown menu labeled "Import from Photo" with a camera icon (`LuCamera`). This would keep the header clean while making the feature findable. No need for a new toolbar button — just one extra menu item would go a long way.

File: packages/ui/src/contexts/AIAnalysisContext.tsx [hunk 4]
```diff
@@ -235,6 +244,48 @@ export const AIAnalysisProvider: React.FC<{ children: React.ReactNode }> = ({ ch
     }
   }, [gameInfo?.komi, analysisCache, updateAnalysisCacheSize, setAnalysisResult, isLoadingSGF]);
 
+  // When numVisits changes, only invalidate the current position's cached result
```
Comment (hadim) [inline packages/ui/src/contexts/AIAnalysisContext.tsx]: 💡 **Suggestion: Consider full cache invalidation on numVisits change**

Currently when `numVisits` changes, only the current position's cache entry is cleared. This means the user sees mixed analysis quality across the game — some positions analyzed with 1 visit, others with 100 visits — which could be confusing, especially when reviewing the analysis graph.

You could consider adding the `numVisits` value to the cache key (so different visit counts coexist) or clearing the entire cache when this setting changes. No strong opinion on which approach — the current behavior is just worth documenting if it's intentional.

**Comment by Aitai** [comment]
Thanks so much for the quick and thoughtful review and for making an exception on the PR size! I really appreciate the leniency. I'll absolutely keep future PRs smaller and more focused. :smile: 

To answer your questions:

The HF model works well as a default as-is. Keeping the standard ONNX means one artifact to maintain, and it runs fine on the WASM backend for devices without GPU support. The in-browser converter handles WebGPU/WebNN on first load and is negligibly fast (54ms on my machine for the 140MB model, so there's no need to ship a pre-converted variant. The flow is: model downloads → auto-converts for the user's GPU backend → result cached for future sessions. Zero configuration required.

The `pytorch_is_available()` check already gates the UI option, so the sidecar is never spawned and the option is never shown to users without Python + PyTorch + GPU. I'll double-check there are no code paths that could crash without Python.

**Comment by hadim** [comment]
> Thanks so much for the quick and thoughtful review and for making an exception on the PR size! I really appreciate the leniency. I'll absolutely keep future PRs smaller and more focused. 😄
> 
> To answer your questions:
> 
> The HF model works well as a default as-is. Keeping the standard ONNX means one artifact to maintain, and it runs fine on the WASM backend for devices without GPU support. The in-browser converter handles WebGPU/WebNN on first load and is negligibly fast (54ms on my machine for the 140MB model, so there's no need to ship a pre-converted variant. The flow is: model downloads → auto-converts for the user's GPU backend → result cached for future sessions. Zero configuration required.
> 
> The `pytorch_is_available()` check already gates the UI option, so the sidecar is never spawned and the option is never shown to users without Python + PyTorch + GPU. I'll double-check there are no code paths that could crash without Python.

Great 👌

File: apps/desktop/src-tauri/src/commands.rs (hunk 1 above)
Comment (Aitai) [reply apps/desktop/src-tauri/src/commands.rs]: Fixed. Added `sanitize_model_id()` that rejects path separators and traversal patterns (`..`, `/`, `\`). Applied to `onnx_get_cached_model`, `onnx_delete_cached_model`, and `save_uploaded_model`. 
For `pytorch_initialize`, the model path is already resolved through our cache system, but I've added a canonicalize + starts_with check against the models directory as an additional safeguard.

File: packages/board-recognition/src/perspective.ts [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+  }
+
+  const h = solveLinear(A, b);
+  if (!h) return null;
+  return [...h, 1]; // append h8=1
+}
+
+/** Apply homography H to point (x, y) → (x', y') */
+export function applyHomography(H: number[], x: number, y: number): [number, number] {
+  const w = H[6] * x + H[7] * y + H[8];
```
[the remark this answers is not in the document]
Comment (Aitai) [reply packages/board-recognition/src/perspective.ts]: Fixed. Added a guard for near-zero `w` that falls back to returning the input point.

**Comment by hadim** [comment]
LGTM. Thanks again 👍
